backend/graph.py

```diff
@@ -0,0 +1,292 @@
+    )
+    weaviate_client = Weaviate(
+        client=weaviate_client,
+        index_name=WEAVIATE_DOCS_INDEX_NAME,
+        text_key="text",
+        embedding=get_embeddings_model(),
+        by_text=False,
+        attributes=["source", "title"],
+    )
+    return weaviate_client.as_retriever(search_kwargs=dict(k=6))
+
+
+def format_docs(docs: Sequence[Document]) -> str:
+    formatted_docs = []
+    for i, doc in enumerate(docs):
+        doc_string = f"<doc id='{i}'>{doc.page_content}</doc>"
+        formatted_docs.append(doc_string)
+    return "\n".join(formatted_docs)
+
+
+def retrieve_documents(state: AgentState):
+    retriever = get_retriever()
+    messages = convert_to_messages(state["messages"])
+    query = messages[-1].content
+    relevant_documents = retriever.get_relevant_documents(query)
+    return {"query": query, "documents": relevant_documents, "messages": []}
+
+
+def retrieve_documents_with_chat_history(state: AgentState, config):
+    model_name = config.get("configurable", {}).get("model_name", OPENAI_MODEL_KEY)
+    retriever = get_retriever()
+    model = get_model(model_name).with_config(tags=["nostream"])
+
+    CONDENSE_QUESTION_PROMPT = PromptTemplate.from_template(REPHRASE_TEMPLATE)
+    condense_question_chain = (
+        CONDENSE_QUESTION_PROMPT | model | StrOutputParser()
+    ).with_config(
+        run_name="CondenseQuestion",
+    )
+
+    messages = convert_to_messages(state["messages"])
+    query = messages[-1].content
+    retriever_with_condensed_question = condense_question_chain | retriever
+    relevant_documents = retriever_with_condensed_question.invoke(
+        {"question": query, "chat_history": get_chat_history(messages)}
+    )
+    return {"query": query, "documents": relevant_documents, "messages": []}
+
+
+def route_to_retriever(
+    state: AgentState,
+) -> Literal["retriever", "retriever_with_chat_history"]:
+    if len(state["messages"]) == 1:
+        return "retriever"
+    else:
+        return "retriever_with_chat_history"
+
+
+def get_chat_history(messages: Sequence[BaseMessage]) -> Sequence[BaseMessage]:
+    chat_history = []
+    for message in messages:
+        if isinstance(message, AIMessage) or isinstance(message, HumanMessage):
+            chat_history.append(message)
+    return chat_history
+
+
+def synthesize_response(
+    state: AgentState, model: LanguageModelLike, prompt_template: str
+):
+    prompt = ChatPromptTemplate.from_messages(
+        [
+            ("system", prompt_template),
+            MessagesPlaceholder(variable_name="chat_history"),
+            ("human", "{question}"),
+        ]
+    )
+    response_synthesizer = prompt | model
+    synthesized_response = response_synthesizer.invoke(
+        {
+            "question": state["query"],
+            "context": format_docs(state["documents"]),
+            "chat_history": get_chat_history(convert_to_messages(state["messages"])),
+        }
+    )
+    return {
+        **state,
```

Comment:
afaiu nodes only need to return patches to state, not the entire updated state

Reply:
ah yes, that's a good point, i'll clean that up